hbase-client/src/main/java/org/apache/hadoop/hbase/client/MasterRegistry.java

```diff
@@ -0,0 +1,364 @@
+import org.apache.hadoop.hbase.shaded.protobuf.generated.MasterProtos.GetMetaRegionLocationsResponse;
+
+/**
+ * Master based registry implementation. Makes RPCs to the configured master addresses from config
+ * {@value HConstants#MASTER_ADDRS_KEY}.
+ *
+ * It has the ability to burst the same RPC to multiple masters as a batch and returns whatever
+ * comes back first (a.k.a hedged RPCs). Number of target masters in a single batch is controlled
+ * via {@value HConstants#MASTER_REGISTRY_NUM_HEDGED_REQS_KEY}. If it is set to 1 (default), it is
+ * equivalent to picking a random master from the configured list.
+ *
+ * TODO: Handle changes to the configuration dynamically without having to restart the client.
+ */
+@InterfaceAudience.Private
+public class MasterRegistry implements AsyncRegistry {
+  private static final Logger LOG = LoggerFactory.getLogger(MasterRegistry.class);
+
+  // Configured list of masters to probe the meta information from.
+  private final List<ServerName> masterServers;
+  // Controls the fan out of the hedged requests. Requests are made in batches of this number until
+  // all the servers are exhausted. The first returned result is passed back to the client.
+  private final int requestFanOut;
+  private ExecutorService masterRpcPool;
+
+  // RPC client used to talk to the masters.
+  private final RpcClient rpcClient;
+  private final RpcControllerFactory rpcControllerFactory;
+  private final int rpcTimeoutNs;
+
+  // A simple interface that callers can implement to make an RPC to master. This is used to
+  // abstract out the logic needed to hedge the requests to multiple masters. For more details, look
+  // at doRPCs().
+  @VisibleForTesting
+  @FunctionalInterface
+  public interface RpcCall<RESP> {
+    RESP doRpc(ClientMetaService.BlockingInterface stub) throws ServiceException;
+  }
+
+  /**
+   * A shared RPC context between a batch of hedged RPCs. Tracks the state and helpers needed to
+   * synchronize on multiple RPCs to different masters fetching the result. All the methods are
+   * thread-safe.
+   * @param <RESP> Return response type for the RPCs.
+   */
+  private class BatchRpcCtx<RESP> {
+    // Result set by the thread finishing first. Set only once.
+    private AtomicReference<RESP> result;
+    // Caller waits on this latch being set.
+    private CountDownLatch resultsReady;
+    // Book-keeping for number of failed RPCs.
+    private AtomicInteger failedRPCs;
+
+    BatchRpcCtx() {
+      result = new AtomicReference<>();
+      // We set this to 1, so that the first successful RPC result is returned to the client.
+      resultsReady = new CountDownLatch(1);
+      failedRPCs = new AtomicInteger(0);
+    }
+
+    /**
+     * Sets the result only if it is not already set by another thread. Thread that successfully
+     * sets the result also count downs the latch.
+     * @param result Result to be set.
+     */
+    public void setResultIfNotSet(RESP result) {
```

Comment:
Return boolean which is true if succeeded in setting result and false otherwise? nit.